src/symmetric_crypto.cpp

```diff
@@ -0,0 +1,349 @@
+/*
+ * #%L
+ * %%
+ * Copyright (C) 2020 BMW Car IT GmbH
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+
+#include "mococrw/symmetric_memory.h"
+
+#include <openssl/evp.h>
+
+#include "mococrw/error.h"
+#include "mococrw/symmetric_crypto.h"
+#include "mococrw/openssl_wrap.h"
+
+namespace mococrw
+{
+
+using namespace openssl;
+
+class AESCipher::Impl
+{
+public:
+    Impl(SymmetricCipherMode mode, SymmetricCipherKeySize keySize, SymmetricCipherPadding padding,
+         const std::vector<uint8_t> &secretKey, const std::vector<uint8_t> &iv, Operation operation,
+         std::unique_ptr<CipherMemoryStrategyI> memoryStrategy = std::make_unique<QueueOfVectorsMemoryStrategy>())
+            : _iv{iv}, _operation{operation}, _mode{mode}, _bufferStrategy(std::move(memoryStrategy))
+    {
+        _ctx = _EVP_CIPHER_CTX_new();
+
+        const EVP_CIPHER *(*EVPCipherConstructor)() = nullptr;
+        switch (mode) {
+            case SymmetricCipherMode::GCM:
+                EVPCipherConstructor = keySize == SymmetricCipherKeySize::S_256  ? EVP_aes_256_gcm : EVP_aes_128_gcm;
```

Comment:
This will silently default to AES-128 when the given parameter is anything but S_256. Is this what we want?

Reply:
There is a check later if key length matches the cipher, but I can make it more explicit.